readme.md
<!--
  :Author: Thiago Navarro
  :Email: thiago@oxyoy.com

  **Created at:** 06/15/2021 12:48:52 Tuesday
  **Modified at:** 09/17/2021 01:03:10 AM Friday

  ------------------------------------------------------------------------------

  readme
  ------------------------------------------------------------------------------
-->

# Formulas

Formulas is a Nim library that helps you calculate areas, volumes, edges, etc...

**Supports version 1.0.0 of Nim!**

Now with configurable PI value!

**More formulas will be added soon**


----
## Tetrahedron

- area
- edge
- height
- volume

Example (`example/tetrahedron.nim`):
```nim
import formulas

var tetrahedron = initTetrahedron(edge = 10)

echo tetrahedron.area() # +- 173.2
echo tetrahedron.height() # +- 8.1
echo tetrahedron.volume() # +- 117.8

tetrahedron.volume = 117.85113019775793 # same as tetrahedron.volume()
tetrahedron.edge = 0

echo tetrahedron.edge() # 10
```

## Cone

- slant height
- side area
- total area
- volume
- radius
- height

Example (`example/cone.nim`):
```nim
import formulas

var cone = initCone(height = 10, radius = 8)

echo cone.slant() # 12.8

cone.slant = cone.slant() # set the slant value to the calculated one

echo cone.volume() # +- 670.2
echo cone.totalArea() # +- 522.9
echo cone.sideArea() # 321.8

cone.volume = cone.volume()
cone.radius = 0

echo cone.radius() # 8

cone.radius = 8
cone.height = 0

echo cone.height() # 10
```

## Docs

The documentation can be found in [Github pages](https://thisago.github.io/formulas/formulas.html)

## TODO

- [ ] Change the nimble package url from gitea to github

## License

MIT


changelog.md
<!--
  Created at: 09/17/2021 00:48:42 Friday
  Modified at: 09/17/2021 12:49:42 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
-->

# Changelog

## Version 0.1.1 (!time)

- Changed comment header in all files
- Added docs


tests/test1.nim
#[
  Created at: 06/15/2021 10:23:57 Tuesday
  Modified at: 09/17/2021 12:50:02 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

import unittest

import ./formulas
from std/math import nil

proc round(x: SomeFloat): int =
  math.round(x).int

suite "Tetrahedron":
  setup:
    var tetrahedron = initTetrahedron(edge = 10)

  test "Area":
    check tetrahedron.area().round == 173
  test "Height":
    tetrahedron.height = 12
    check tetrahedron.height().round == 8
  test "Volume":
    tetrahedron.volume = 10
    check tetrahedron.volume().round == 118
  test "Edge":
    tetrahedron.edge = 0
    tetrahedron.volume = 4.23
    check tetrahedron.edge().round == 3

suite "Cone":
  setup:
    var cone = initCone(height = 10, radius = 8)

  test "Slant height":
    check cone.slant().round == 13
  test "Side area":
    cone.slant = cone.slant()
    check cone.sideArea().round == 322
  test "Total area":
    cone.slant = cone.slant()
    check cone.totalArea().round == 523
  test "Volume":
    check cone.volume().round == 670
  test "Radius":
    cone.volume = 670.2
    cone.radius = 0
    check cone.radius().round == 8
  test "Height":
    cone.volume = 670.2
    cone.radius = 8
    cone.height = 0
    check cone.height().round == 10


src/formulas/tetrahedron.nim
#[
  Created at: 06/14/2021 10:35:59 Monday
  Modified at: 09/17/2021 12:51:28 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

##[
  Tetrahedron Formulas

  https://www.matematica.pt/en/useful/math-formulas.php
]##

import math

import ./base

# Tetrahedron

type
  Tetrahedron = object of Shape
    areaVal*, edgeVal*, heightVal*, volumeVal*: float

# Init

func initTetrahedron*(area, edge, height, volume: SomeNumber = 0): Tetrahedron =
  ## Creates new tetrahedron
  Tetrahedron(
    areaVal: float area,
    edgeVal: float edge,
    heightVal: float height,
    volumeVal: float volume
  )

# Setters

func `area=`*(self: var Tetrahedron, value: SomeNumber) = self.areaVal = float value
func `edge=`*(self: var Tetrahedron, value: SomeNumber) = self.edgeVal = float value
func `height=`*(self: var Tetrahedron, value: SomeNumber) = self.heightVal = float value
func `volume=`*(self: var Tetrahedron, value: SomeNumber) = self.volumeVal = float value

# Calculations

func area*(self: Tetrahedron): float =
  ## Calculates the tetrahedron area with edge
  assert self.edgeVal != 0, "No edge provided"
  pow(self.edgeVal, 2) * sqrt 3.0
func height*(self: Tetrahedron): float =
  ## Calculates the tetrahedron height with edge
  assert self.edgeVal != 0, "No edge provided"
  (self.edgeVal * sqrt 6.0) / 3
func volume*(self: Tetrahedron): float =
  ## Calculates the tetrahedron volume (cm3) with edge
  assert self.edgeVal != 0, "No edge provided"
  (pow(self.edgeVal, 3) * sqrt 2.0) / 12
func edge*(self: Tetrahedron): float =
  ## Calculates the tetrahedron edge with volume (cm3)
  assert self.volumeVal != 0, "No volume provided"
  cbrt((self.volumeVal * 12) / sqrt 2.0)


when isMainModule:
  var tetrahedron = initTetrahedron(edge = 10)
  echo tetrahedron.area

  tetrahedron.edge = 12.0
  echo tetrahedron.height

  tetrahedron.edge = 10
  echo tetrahedron.volume

  tetrahedron.edge = 0
  tetrahedron.volume = 4.23
  echo tetrahedron.edge


src/formulas/triangle.nim
#[
  Created at: 06/16/2021 09:57:16 Wednesday
  Modified at: 09/17/2021 12:58:19 AM Friday

        Copyright (C) 2021 Thiago Navarro
  See file "license" for details about copyright
]#

##[
  **WIP**

  Triangle Formulas

  https://en.wikipedia.org/wiki/Triangle
]##

import ./base
import math

type
  Triangle = object of Shape
    baseEdgeVal*, edgeVal*, parallelSectionVal*, heightVal*: float

# Init

func initTriangle*(baseEdge, edge, parallelSection, height: SomeNumber = 0): Triangle =
  ## Creates new Triangle
  Triangle(
    baseEdgeVal: float baseEdge,
    edgeVal: float edge,
    parallelSectionVal: float parallelSection,
    heightVal: float height,
  )

# Setters

func `baseEdge=`*(self: var Triangle, value: SomeNumber) = self.baseEdgeVal = float value
func `edge=`*(self: var Triangle, value: SomeNumber) = self.edgeVal = float value
func `parallelSection=`*(self: var Triangle, value: SomeNumber) = self.parallelSectionVal = float value
func `height=`*(self: var Triangle, value: SomeNumber) = self.heightVal = float value

# Calculations

func slant*(self: Triangle): float =
  ## Calculates the cone slant height with height and radius
  doAssert self.heightVal != 0, "No height provided"
  doAssert self.radiusVal != 0, "No radius provided"
  sqrt(pow(self.heightVal, 2) + pow(self.radiusVal, 2))

func sideArea*(self: Triangle): float =
  ## Calculates the cone side area (cm2) with slant height and radius
  doAssert self.radiusVal != 0, "No radius provided"
  doAssert self.slantVal != 0, "No slant provided"
  self.pi * self.radiusVal * self.slantVal

func totalArea*(self: Triangle): float =
  ## Calculates the cone total area (cm2) with slant height and radius
  doAssert self.radiusVal != 0, "No radius provided"
  doAssert self.slantVal != 0, "No slant provided"
  self.pi * self.radiusVal * (self.radiusVal + self.slantVal)

func volume*(self: Triangle): float =
  ## Calculates the cone volume with height and radius
  doAssert self.heightVal != 0, "No height provided"
  doAssert self.radiusVal != 0, "No radius provided"
  (self.pi * pow(self.radiusVal, 2) * self.heightVal) / 3

func radius*(self: Triangle): float =
  ## Calculates the cone radius with height and volume
  doAssert self.heightVal != 0, "No height provided"
  doAssert self.volumeVal != 0, "No volume provided"
  sqrt((3 * self.volumeVal) / (self.pi * self.heightVal))

func height*(self: Triangle): float =
  ## Calculates the cone height with height and volume
  doAssert self.radiusVal != 0, "No radius provided"
  doAssert self.volumeVal != 0, "No volume provided"
  (self.volumeVal * 3) / (self.pi * pow(self.radiusVal, 2))


when isMainModule:
  var cone = initTriangle(height = 10, radius = 8)

  echo cone.slant # 12.8
  echo cone.sideArea # 321.6
  echo cone.totalArea # 522.7
  echo cone.volume # 670.2

  cone.pi = 3.14

  echo cone.slant # 12.8
  echo cone.sideArea # 321.6
  echo cone.totalArea # 522.7
  echo cone.volume # 670.2

  cone.volume = 670.2064327658226
  cone.radius = 0
  echo cone.radius # 8

  cone.volume = 670.2064327658226
  cone.radius = 8
  cone.height = 0
  echo cone.height # 10
